src/v2/providers/eventarc.ts

```diff
@@ -0,0 +1,163 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2017 Firebase
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import * as options from '../options';
+import { CloudEvent, CloudFunction } from '../core';
+import { copyIfPresent } from '../../common/encoding';
+import { ManifestEndpoint } from '../../runtime/manifest';
+
+/** Options that can be set on an eventarc trigger. */
+export interface EventarcTriggerOptions extends options.EventHandlerOptions {
+  /**
+   * ID of the channel. Can be either:
+   *   * fully qualified channel resource name:
+   *     `projects/{project}/locations/{location}/channels/{channel-id}`
+   *   * partial resource name with location and channel id, in which case
+   *     the runtime project ID of the function will be used:
+   *     `locations/{location}/channels/{channel-id}`
+   *   * partial channel-id, in which case the runtime project ID of the
+   *     function and `us-central1` as location will be used:
+   *     `{channel-id}`
+   * 
+   * If not specified, the default Firebase channel will be used:
+   * `projects/{project}/locations/us-central1/channels/firebase`
+   */
+   channel?: string;
+
+  /**
+   * Eventarc event exact match filter.
+   */
+   filters?: Record<string, string>;
+ }
+
+export type CloudEventHandler = (event: CloudEvent<any>) => any | Promise<any>;
+
+/** Handle an Eventarc event published on the default channel. */
+export function onCustomEventPublished<T = any>(
+  eventType: string,
+  handler: CloudEventHandler
+): CloudFunction<CloudEvent<T>>;
+
+export function onCustomEventPublished<T = any>(
+  eventType: string,
+  opts: EventarcTriggerOptions,
+  handler: CloudEventHandler
+): CloudFunction<CloudEvent<T>>;
+
+export function onCustomEventPublished<T = any>(
+  eventType: string,
+  optsOrHandler:
```

Comment:
This is a BAD time to discuss API design, but this seems divergent from existing v2 APIs.

e.g. for storage we have:

```
const fn = onObjectArchived(() => 42);
const fnWithBucket = onObjectArchived("some-bucket", () => 43);
const fnWithOpts = onObjectArchived({ bucket: "some-bucket" }, () => 43);
```

Then I'd expect custom event published API to support following signatures:

```
const fn = onCustomEventPublished("event-type", () => 42);
const fnWithOptions = onCustomEventPublished({ eventType: "event-type" }, () => 42)
```

This is also consistent with recently added FireAlerts.

@colerogers @inlined @pavelgj  Should we file an addendum API proposal or is there a reason for an exception here?

Reply:
Yeah, I'd rather not reopen this can... there was a LOT of back and forth in go/rvtfmkswhlks and the road to this API was long.

In short, the difference from storage where bucket is optional (will use default bucket if not specified) for custom events type is never optional, it must always be specified, which is hard to express in the API with your proposal.

Reply:
sgtm

Reply:
@pavelgj @taeold  Here's an example of the alerts interface - https://github.com/firebase/firebase-functions/blob/6d146aa5604c899f21b19a749b8e6663bddef697/src/v2/providers/alerts/alerts.ts#L56

The user must give an `alertType` parameter, they can either use a string or the options object (the options object requires a string). I think this fits your `type` parameter use case.

Reply:
Done. Amended API proposal: go/ryvvt